File: 0328-奇偶链表/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        odd_head, even_head = ListNode(0), ListNode(0)
        odd_tail, even_tail = odd_head, even_head
        flag = True
        while head:
            if flag:
                odd_tail.next = head
                odd_tail = odd_tail.next
            else:
                even_tail.next = head
                even_tail = even_tail.next
            flag = not flag
            head = head.next
        even_tail.next = None
        odd_tail.next = even_head.next
        return odd_head.next

File: 0328-奇偶链表/Solution.java
//Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode oddTail = head;
        ListNode evenTail = head.next;
        ListNode work = head.next.next;
        boolean odd = true;
        while (work != null) {
            if (odd) {
                ListNode tmp = work;
                evenTail.next = tmp.next;
                work = tmp.next;
                tmp.next = oddTail.next;
                oddTail.next = tmp;
                oddTail = tmp;
            } else {
                evenTail = work;
                work = work.next;
            }
            odd = !odd;
        }
        return head;
    }
}